### PR TITLE
[GDBJIT] Display DynaRec info in source file

### PR DESCRIPTION
**Pull request by ksco**
gdbjit showcase

### CODE REVIEW

File: src/dynarec/arm64/dynarec_arm64_functions.c [hunk 1]
```diff
@@ -661,97 +661,112 @@ const char* getCacheName(int t, int n)
...
+        if (dyn->insts[ninst].nat_flags_op == NAT_FLAG_OP_TOUCH && dyn->insts[ninst].before_nat_flags)
+            length += sprintf(buf + length, " NF:%d/read:%x", dyn->insts[ninst].nat_flags_op, dyn->insts[ninst].before_nat_flags);
+        else
+            length += sprintf(buf + length, " NF:%d", dyn->insts[ninst].nat_flags_op);
+    }
+    if (dyn->insts[ninst].use_nat_flags || dyn->insts[ninst].set_nat_flags || dyn->insts[ninst].need_nat_flags) {
+        length += sprintf(buf + length, " nf:%hhx/%hhx/%hhx", dyn->insts[ninst].set_nat_flags, dyn->insts[ninst].use_nat_flags, dyn->insts[ninst].need_nat_flags);
+    }
+    if (dyn->insts[ninst].invert_carry)
+        length += sprintf(buf + length, "CI");
```
Comment (ptitSeb) [inline src/dynarec/arm64/dynarec_arm64_functions.c:693]: missing space in " CI"

File: src/dynarec/arm64/dynarec_arm64_functions.c [hunk 2]
```diff
@@ -661,97 +661,112 @@ const char* getCacheName(int t, int n)
...
+    if (dyn->insts[ninst].gen_inverted_carry)
+        length += sprintf(buf + length, "gic");
```
Comment (ptitSeb) [inline src/dynarec/arm64/dynarec_arm64_functions.c:695]: missing space in " gic"

File: src/dynarec/arm64/dynarec_arm64_functions.c [hunk 3]
```diff
@@ -661,97 +661,112 @@ const char* getCacheName(int t, int n)
...
-                case NEON_CACHE_YMMW: dynarec_log(LOG_NONE, " Q%d:%s", ii, getCacheName(dyn->insts[ninst].n.neoncache[ii].t, dyn->insts[ninst].n.neoncache[ii].n)); break;
-                case NEON_CACHE_YMMR: dynarec_log(LOG_NONE, " Q%d:%s", ii, getCacheName(dyn->insts[ninst].n.neoncache[ii].t, dyn->insts[ninst].n.neoncache[ii].n)); break;
-                //case NEON_CACHE_SCR: dynarec_log(LOG_NONE, " D%d:%s", ii, getCacheName(dyn->insts[ninst].n.neoncache[ii].t, dyn->insts[ninst].n.neoncache[ii].n)); break;
-                case NEON_CACHE_NONE:
-                default:    break;
-            }
+    if (!box64_dynarec_dump && !box64_dynarec_gdbjit) return;
+
+    static char buf[512];
+    int length = sprintf(buf, "barrier=%d state=%d/%d/%d(%d:%d->%d:%d), %s=%X/%X, use=%X, need=%X/%X, sm=%d(%d/%d)",
```
Comment (ptitSeb) [inline src/dynarec/arm64/dynarec_arm64_functions.c:667]: were are the colors for regular dump?

File: src/dynarec/arm64/dynarec_arm64_functions.c [hunk 4]
```diff
@@ -661,97 +661,112 @@ const char* getCacheName(int t, int n)
...
+        length += sprintf(buf + length, " X87:%d/%d(+%d/-%d)%d", dyn->n.stack, dyn->insts[ninst].n.stack_next, dyn->insts[ninst].n.stack_push, dyn->insts[ninst].n.stack_pop, dyn->insts[ninst].n.x87stack);
+    }
+    if (dyn->insts[ninst].n.combined1 || dyn->insts[ninst].n.combined2) {
+        length += sprintf(buf + length, " %s:%d/%d", dyn->insts[ninst].n.swapped ? "SWP" : "CMB", dyn->insts[ninst].n.combined1, dyn->insts[ninst].n.combined2);
+    }
+    if (box64_dynarec_dump) {
+        printf_x64_instruction(rex.is32bits ? my_context->dec32 : my_context->dec, &dyn->insts[ninst].x64, name);
+        dynarec_log(LOG_NONE, "%s%p: %d emitted opcodes, inst=%d, %s%s\n",
+            (box64_dynarec_dump > 1) ? "\e[32m" : "",
+            (void*)(dyn->native_start + dyn->insts[ninst].address), dyn->insts[ninst].size / 4, ninst, buf, (box64_dynarec_dump > 1) ? "\e[m" : "");
```
Comment (ksco) [inline src/dynarec/arm64/dynarec_arm64_functions.c:762]: @ptitSeb Colors are kept here.

File: src/dynarec/arm64/dynarec_arm64_functions.c [hunk 5]
```diff
@@ -661,97 +661,112 @@ const char* getCacheName(int t, int n)
...
+        for (int ii = 0; ii < dyn->insts[ninst].pred_sz; ++ii)
+            length += sprintf(buf + length, "%s%d", ii ? "/" : "", dyn->insts[ninst].pred[ii]);
+    }
+    if (!dyn->insts[ninst].x64.alive)
+        length += sprintf(buf + length, "not executed");
+    if (dyn->insts[ninst].x64.jmp && dyn->insts[ninst].x64.jmp_insts >= 0) {
+        length += sprintf(buf + length, ", jmp=%d", dyn->insts[ninst].x64.jmp_insts);
+    }
+    if (dyn->insts[ninst].x64.jmp && dyn->insts[ninst].x64.jmp_insts == -1)
+        length += sprintf(buf + length, "jmp=out");
```
Comment (ptitSeb) [inline src/dynarec/arm64/dynarec_arm64_functions.c:713]: missing comma in ", jmp-out"

File: src/dynarec/arm64/dynarec_arm64_functions.c [hunk 6]
```diff
@@ -661,97 +661,112 @@ const char* getCacheName(int t, int n)
...
+            default: break;
         }
-        if(memcmp(dyn->insts[ninst].n.neoncache, dyn->n.neoncache, sizeof(dyn->n.neoncache))) {
-            dynarec_log(LOG_NONE, " %s(Change:", (box64_dynarec_dump>1)?"\e[1;91m":"");
-            for(int ii=0; ii<32; ++ii) if(dyn->insts[ninst].n.neoncache[ii].v!=dyn->n.neoncache[ii].v) {
-                dynarec_log(LOG_NONE, " V%d:%s", ii, getCacheName(dyn->n.neoncache[ii].t, dyn->n.neoncache[ii].n));
-                dynarec_log(LOG_NONE, "->%s", getCacheName(dyn->insts[ninst].n.neoncache[ii].t, dyn->insts[ninst].n.neoncache[ii].n));
+    }
+    if (memcmp(dyn->insts[ninst].n.neoncache, dyn->n.neoncache, sizeof(dyn->n.neoncache))) {
+        length += sprintf(buf + length, " %s(Change:", (box64_dynarec_dump > 1) ? "\e[1;91m" : "");
```
Comment (ptitSeb) [inline src/dynarec/arm64/dynarec_arm64_functions.c:735]: you get color here, even for gdbjit stuffs. Is that ok?

File: src/dynarec/arm64/dynarec_arm64_functions.c [hunk 7]
```diff
@@ -661,97 +661,112 @@ const char* getCacheName(int t, int n)
...
-        if(dyn->ymm_zero || dyn->insts[ninst].ymm0_add || dyn->insts[ninst].ymm0_sub || dyn->insts[ninst].ymm0_out)
-            dynarec_log(LOG_NONE, " ymm0=(%04x/%04x+%04x-%04x=%04x)", dyn->ymm_zero, dyn->insts[ninst].ymm0_in, dyn->insts[ninst].ymm0_add ,dyn->insts[ninst].ymm0_sub, dyn->insts[ninst].ymm0_out);
-        if(dyn->insts[ninst].purge_ymm)
-            dynarec_log(LOG_NONE, " purgeYmm=%04x", dyn->insts[ninst].purge_ymm);
-        if(dyn->n.stack || dyn->insts[ninst].n.stack_next || dyn->insts[ninst].n.x87stack)
-            dynarec_log(LOG_NONE, " X87:%d/%d(+%d/-%d)%d", dyn->n.stack, dyn->insts[ninst].n.stack_next, dyn->insts[ninst].n.stack_push, dyn->insts[ninst].n.stack_pop, dyn->insts[ninst].n.x87stack);
-        if(dyn->insts[ninst].n.combined1 || dyn->insts[ninst].n.combined2)
-            dynarec_log(LOG_NONE, " %s:%d/%d", dyn->insts[ninst].n.swapped?"SWP":"CMB", dyn->insts[ninst].n.combined1, dyn->insts[ninst].n.combined2);
-        dynarec_log(LOG_NONE, "%s\n", (box64_dynarec_dump>1)?"\e[m":"");
+        length += sprintf(buf + length, ")%s", (box64_dynarec_dump > 1) ? "\e[0;32m" : "");
```
Comment (ptitSeb) [inline src/dynarec/arm64/dynarec_arm64_functions.c:741]: same, color attribute again here.

File: src/dynarec/arm64/dynarec_arm64_functions.c (hunk 3 above)
Comment (ksco) [reply src/dynarec/arm64/dynarec_arm64_functions.c:667]: It's kept down below, see https://github.com/ptitSeb/box64/pull/2179#discussion_r1893794006

File: src/dynarec/arm64/dynarec_arm64_functions.c (hunk 6 above)
Comment (ksco) [reply src/dynarec/arm64/dynarec_arm64_functions.c:735]: Unless one uses DUMP=2 and GDBJIT=1 at the same time, which does not make much sense, there won't be any color, not sure if there is a better way to do this.

**Pull request merged**